server/src/main/java/org/apache/druid/server/coordinator/helper/DruidCoordinatorRuleRunner.java

```diff
@@ -84,8 +85,10 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
     // find available segments which are not overshadowed by other segments in DB
     // only those would need to be loaded/dropped
     // anything overshadowed by served segments is dropped automatically by DruidCoordinatorCleanupOvershadowed
-    final Set<DataSegment> overshadowed = ImmutableDruidDataSource
-        .determineOvershadowedSegments(params.getAvailableSegments());
+    // If metadata store hasn't been polled yet, use empty overshadowed list
+    final Set<SegmentId> overshadowed = Optional
+        .ofNullable(coordinator.getMetadataSegmentManager().getOvershadowedSegments())
```

Comment:
`DruidCoordinator` and its helpers such as `DruidCoordinatorSegmentInfoLoader` should fetch the current DatasourcesSnapshot from `MetadataSegmentManager` at the start of the coordinator loop and then operate on that snapshot exclusively in the rest of the coordinator loop.

Reply:
Added a `getDataSourcesSnapshot()` in `MetadataSegmentManager`. `DruidCoordinator` uses it to get a snapshot of dataSources and operates on that at rest of the places.